src/loaders/zarr/pixel-source.ts

```diff
@@ -0,0 +1,115 @@
+import { BoundsCheckError } from 'zarr';
+import { isInterleaved } from '../utils';
+import { getIndexer } from './lib/indexer';
+
+import type { HTTPStore } from './lib/storage';
+import type { ZarrArray } from 'zarr';
+import type { RawArray } from 'zarr/dist/types/rawArray';
+import type { AsyncStore } from 'zarr/dist/types/storage/types';
+
+import type {
+  PixelSource,
+  Labels,
+  RasterSelection,
+  PixelSourceSelection,
+  LayerData,
+  TileSelection
+} from '../../types';
+
+const DTYPE_LOOKUP = {
+  u1: 'Uint8',
+  u2: 'Uint16',
+  u4: 'Uint32',
+  f4: 'Float32'
+} as const;
+
+type ZarrIndexer<S extends string[]> = (
+  sel: { [K in S[number]]: number } | number[]
+) => number[];
+
+interface ZarrTileSelection {
+  x: number;
+  y: number;
+  selection: number[];
+  signal?: AbortSignal;
+}
+
+class ZarrPixelSource<S extends string[]> implements PixelSource<S> {
+  private _data: ZarrArray;
+  private _indexer: ZarrIndexer<S>;
+
+  constructor(data: ZarrArray, public labels: Labels<S>) {
+    this._indexer = getIndexer(labels);
+    this._data = data;
+  }
+
+  get tileSize() {
+    return this._data.chunks[this._xIndex];
+  }
+
+  get shape() {
+    return this._data.shape;
+  }
+
+  get dtype() {
+    const suffix = this._data.dtype.slice(1) as keyof typeof DTYPE_LOOKUP;
+    if (!(suffix in DTYPE_LOOKUP)) {
+      throw Error(`Zarr dtype not supported, got ${suffix}.`);
+    }
+    return DTYPE_LOOKUP[suffix];
+  }
+
+  private get _xIndex() {
+    const interleave = isInterleaved(this._data.shape);
+    return this._data.shape.length - (interleave ? 2 : 1);
+  }
+
+  private _chunkIndex<T>(
+    selection: PixelSourceSelection<S> | number[],
+    x: T,
+    y: T
+  ) {
+    const sel: (number | T)[] = this._indexer(selection);
+    sel[this._xIndex] = x;
+    sel[this._xIndex - 1] = y;
+    return sel;
+  }
+
+  async getRaster({ selection }: RasterSelection<S> | { selection: number[] }) {
+    const sel = this._chunkIndex(selection, null, null);
+    const { data, shape } = (await this._data.getRaw(sel)) as RawArray;
+    const [height, width] = shape;
+    return { data, width, height } as LayerData;
+  }
+
+  async getTile(props: TileSelection<S> | ZarrTileSelection) {
+    const { x, y, selection, signal } = props;
+    const sel = this._chunkIndex(selection, x, y);
+
+    const store = this._data.store as HTTPStore | AsyncStore<ArrayBuffer>;
+
+    // Injects `signal` prior to zarr.js calling `store.getItem`
+    if (signal && '__vivAddSignal' in store) {
+      store.__vivAddSignal(signal);
```

Comment:
Is there concern about race conditions here since this could be called multiple times before any of the calling tiles return/are cancelled?  So one tile cancelling another's request is my concern.

Reply:
I can't think of a race condition. If the signal is aborted for _any_ of the loaderseleciton requests, it throws and bubbles up within `getTileData` there is no such thing as a "returned tile" it gets aborted.

Reply:
I guess I'm wondering if this is possible:
1. getTile is called for {x: 0, y: 0, sel} and sets the signal on the store
2. getTile is called for {x: 0, y: 1, sel} and sets the signal on the store (overriding the old signal)
3. {x: 0, y: 0, sel} is cancelled
4. getRawChunk is called for {x: 0, y: 0, sel}
But then getRawChunk is using the signal of {x: 0, y: 1, sel} so it doesn't see that {x: 0, y: 0, sel} has been cancelled

Reply:
In the context of `getTileData` there is a single `AbortSingal` provided by deck.gl. So if there was some type of overriding behavior it would all be the same `signal`, correct?  However, I don't think this can happen.

We use `Promise.all` in a `try/catch`within `getTileData`. `Promise.all` returns a _single_ Promise that either resolves to an array of fulfilled promises or ["rejects immediately upon any of the input promises rejecting or non-promises throwing an error, and will reject with this first rejection message / error"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all). 

I guess an alternative would be for `getTIleData` to set the signal before `Promise.all` and then we handle clearing the signal in a `finally` block. But this approach would require that `MultiScaleImageLayer` is aware of a special zarr source.

Reply:
> In the context of `getTileData` there is a single `AbortSingal` provided by deck.gl. So if there was some type of overriding behavior it would all be the same `signal`, correct? However, I don't think this can happen.

One `AbortSignal` per call to `getTileData`, not one for all calls. So each tile gets its own signal (my example differed in tile, not selection for this reason).  Within a single `x,y,z` for a tile, the selections will share the signal but it doesn't seem like the signal should be global to the store, which is used repeatedly for every request.

Reply:
Hmm I'll have to think about that. The idea is to set `HTTPStore._signal` right before calling `getItem` and immediately set `HTTPStore._signal` to `undefined` after. To be conservative, within the `catch` block of `getTileData` we can again set singal to `undefined` for the store to ensure it's removed. 

An important point is that `getTileData` is called concurrently, not in parallel. I've played around with this a bit. You can add a log within the `catch` block of `getTileData` and see requests being cancelled. This approach is far more simple than what we had before (which duplicated a lot of low-level zarr.js _and_ did so improperly). 

If there are noticeable issues, I think we can think about alternative solutions (e.g. making a PR to zarr.js).

Reply:
I opened a PR in Zarr.js to add the ability to pass a signal. I think this implementation is reasonable for now, and if/when that is merged we can remove _and_ use the `HTTPStore` from the new version of zarr.